File: 21. Merge Two Sorted Lists/CnE.py
'''
21. Merge Two Sorted Lists
https://leetcode.com/problems/merge-two-sorted-lists/description/
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        nn = result = ListNode() 
        # Get an empty node to start
        # Keep one pointer always at the starting

        while list1 and list2:
            # When both lists are not empty
            if list1.val < list2.val:
                # Simultaneous assignment
                # For the list and new node (nn) pointer
                list1, nn.next = list1.next, list1
            else:
                # Otherwise assignment for list 2
                list2, nn.next = list2.next, list2
            nn = nn.next
            # Move the new node pointer to the next position
            # For new assignments

        nn.next = list1 if list1 else list2
        # Assign the reset list (could be an empty list)

        return result.next
        # Return the result.next to ignore the initial node
            

File: 21. Merge Two Sorted Lists/CnE.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = (list1, list2) => {

    var initialNode = curr = {var:0, next: null}
    // Get an empty node to start

    while (list1 && list2) {
        // When both lists are not empty
        list1.val < list2.val ? [curr.next, list1] = [list1, list1.next] : [curr.next, list2] = [list2, list2.next]
        // Simutaneous assignment according to value (val)
        curr = curr.next
        // Move the new node pointer to the next position
        // For new assignments
    }
    curr.next = list1 || list2
    // Assign the reset list (could be an empty list)
    return initialNode.next
    // Return the result.next to ignore the initial node
};